### ISSUE TITLE
spike: what is the cause of the chapcs comm=ofi full testing timeouts?

### ISSUE THREAD

**Issue by gbtitus** [chapel-lang/chapel#12197]
Comm=ofi full testing with the sockets provider resulted in about 100 timeouts.  (See #11856, #11942, #12064.)  The immediate cause of these is pretty well understood; some PUTs are much slower than with the sockets provider on XC.  But why?  Although the sockets provider isn't one we expect to use in production we still need to understand this, because the `many-to-one` test doesn't demonstrate any significant difference in sockets performance between XC and chapcs, yet here we have about 100 benchmarks (standing in for applications) that are much slower.

**Comment by gbtitus** [comment]
I added a second receive context to the ofi comm layer, and a completion counter and address vector to go with it, and moved everything but the AM request traffic onto that.  So basically, RMA and AMO traffic is all on the new one.  Progress on this new rx ctx is achieved by having everyone who does RMA or AMOs read the counter as they wait for completion, plus there's backup from the AM handler which checks it at intervals as it waits for requests.

I re-ran the tests that timed out in the original full testing, and this does seem to have helped.  The testing did run over its slurm job time limit and get killed partway through, but based on the results this change does seem to have helped.  The run attempted 42 tests.  10 of these timed out, and an additional 4 ran for close enough to 300 seconds that it may have been just luck (or load on chapcs) that they timed out the first time and passed this time.  But the other 32 finished comfortably under the timeout limit, in many cases in single- or low double-digits of seconds.

Next I'm going to look at some of the tests that timed out in this 2nd run, in hopes of finding another problem.

**Comment by gbtitus** [comment]
After looking more deeply at the timeouts from the re-run and adding some more instrumentation, it appeared that while moving the RMA and AMOs to a different endpoint helped, sharing the progress load for that endpoint across all the non-AM handler threads was probably leading to cache thrashing and thus timeouts, in programs that communicated a lot.   As I thought about solutions for this, everything I came up with ended up looking effectively like limiting progress responsibility to one thread at a time, while switching that responsibility among threads at a fairly slow rate.  Ultimately I decided that especially for a provider like sockets whose purpose is complete functionality rather than performance, it was better just to use automatic progress.  So I re-coded to default to automatic progress with the sockets provider and manual progress otherwise, and re-ran.  Gratifyingly, along with some other changes I'd made to avoid threads monopolizing cores while waiting for network transactions and AMs to finish, this solved all the timeouts except for these:

- library/packages/ZMQ/multilocale/hello-multlocale
- library/packages/ZMQ/multilocale/req-rep
- studies/ssca2/rachels/SSCA2_test (compopts: 1-5)

I'm re-running also on XC (with sockets), and it looks like nearly everything sped up there as well.

In addition, this improves the many-to-one performance, especially for op=on and op=fastOn, though more on chapcs than on XC.

Next steps: there are now about 20 tests that fail for various reasons, including the above.  I'm going to make a pass through those and see how many I can clean up in one more day.
